internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/PubSubSharedProducerAdapter.java

```diff
@@ -0,0 +1,178 @@
+    long startNs = System.nanoTime();
+    Future<PubSubProduceResult> result = producerAdapter.sendMessage(topic, partition, key, value, headers, callback);
+    sharedProducerStats.recordProducerSendLatency(LatencyUtils.getLatencyInMS(startNs));
+    return result;
+  }
+
+  @Override
+  public void flush() {
+    producerAdapter.flush();
+  }
+
+  @Override
+  public void close(int closeTimeOutMs) {
+    producerAdapter.close(closeTimeOutMs);
+  }
+
+  @Override
+  public void close(int closeTimeOutMs, boolean doFlush) {
+    producerAdapter.close(closeTimeOutMs, doFlush);
+  }
+
+  @Override
+  public void close(String topic, int closeTimeoutMs) {
+    if (sharedProducerFactory.isRunning()) {
+      sharedProducerFactory.releaseSharedProducer(topic);
+    } else {
+      LOGGER.info("Producer is already closed, can't release for topic: {}", topic);
+    }
+  }
+
+  @Override
+  public void close(String topic, int closeTimeoutMs, boolean doFlush) {
+    if (doFlush) {
+      producerAdapter.flush();
+    }
+    close(topic, closeTimeoutMs);
+  }
+
+  @Override
+  public Object2DoubleMap<String> getMeasurableProducerMetrics() {
+    return producerAdapter.getMeasurableProducerMetrics();
+  }
+
+  @Override
+  public String getBrokerAddress() {
+    return producerAdapter.getBrokerAddress();
+  }
+
+  public int getId() {
+    return id;
+  }
+
+  public synchronized void addProducerTask(String producerTaskName) {
+    producerTasks.add(producerTaskName);
+  }
+
+  public synchronized void removeProducerTask(String producerTaskName) {
+    producerTasks.remove(producerTaskName);
+  }
+
+  public int getProducerTaskCount() {
+    return producerTasks.size();
+  }
+
+  public String toString() {
+    return "{Id: " + id + ", Task Count: " + getProducerTaskCount() + "}";
+  }
+
+  private synchronized void mayBeCalculateAllProducerMetrics() {
+    if (LatencyUtils.getElapsedTimeInMs(lastStatUpdateTsMs) < 60 * Time.MS_PER_SECOND) {
+      return;
+    }
+
+    // measure
+    Object2DoubleMap<String> metrics = producerAdapter.getMeasurableProducerMetrics();
+    producerMetrics.replaceAll((n, v) -> metrics.getOrDefault(n, StatsErrorCode.KAFKA_CLIENT_METRICS_DEFAULT.code));
+    lastStatUpdateTsMs = System.currentTimeMillis();
+  }
+
+  private class SharedProducerStats extends AbstractVeniceStats {
+    private final Sensor producerSendLatencySensor;
+
+    public SharedProducerStats(MetricsRepository metricsRepository) {
+      super(metricsRepository, "PubSubSharedProducer");
+      producerMetrics.keySet().forEach(metric -> {
+        String metricName = "producer_" + id + "_" + metric;
+        LOGGER.info("Registering metric: {}", metricName);
+        registerSensorIfAbsent(metricName, new Gauge(() -> {
+          mayBeCalculateAllProducerMetrics();
+          return producerMetrics.get(metric);
+        }));
+      });
+      producerSendLatencySensor = registerSensor("producer_" + id + "_send_api_latency", new Avg(), new Max());
```

Comment:
nit: maybe we could remove "_api_" here and match pattern with other sensor/metric naming.

Reply:
Done